src/ServiceControl.Monitoring/Diagrams/MonitoringData.cs

```diff
@@ -0,0 +1,55 @@
+﻿namespace ServiceControl.Monitoring.Raw
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+
+    public class MonitoringData
+    {
+        public MonitoringData(int historySize)
+        {
+            this.historySize = historySize;
+
+            Endpoints = new ConcurrentDictionary<string, EndpointData>();
+        }
+
+        public EndpointData Get(string endpointName)
+        {
+            return Endpoints.GetOrAdd(endpointName, new EndpointData(historySize));
+        }
+
+        public ConcurrentDictionary<string, EndpointData> Endpoints;
+        readonly int historySize;
+    }
+
+    public class EndpointData
+    {
+        public EndpointData(int size)
+        {
+            Timestamps = new DateTime[size];
+            CriticalTime = new float?[size];
+            ProcessingTime = new float?[size];
+
+            this.size = size;
+            head = 0;
+        }
+
+        public void Record(DateTime timestamp, float? criticalTime, float? processingTime)
+        {
+            Interlocked.Increment(ref head);
+
+            var index = (head - 1) % size;
+
+            Timestamps[index] = timestamp;
+            CriticalTime[index] = criticalTime;
+            ProcessingTime[index] = processingTime;
```

Comment:
All the values are written with no memory fence. Effectively their assignments might be postponed in time as the reader sets no barrier as well. Instead of having a class with 3 arrays, we could think about having an array of structs with three fields `Timestamp`, `CriticalTime`, `ProcessingTime` and one additional Status for a barrier. When writing under specific index, you could assign all the values and `Volatile.Write (ref records[index].Barrier, 1)` the barrier.
When reading, first the `Volatile.Read (ref records[index].Barrier)` would be called with followed reads of the rest of items.
It might be not needed if the visibility of these values is ensured in another way (but I see nothing like this).